snuba/clickhouse/formatter/query.py

```diff
@@ -1,174 +1,251 @@
 from dataclasses import replace
-from typing import Optional
+from functools import partial
+from typing import Callable, Optional, Sequence, Union
 
 from snuba import settings as snuba_settings
 from snuba.clickhouse.formatter.expression import ClickhouseExpressionFormatter
 from snuba.clickhouse.formatter.nodes import (
     FormattedNode,
     FormattedQuery,
+    FormattedSubQuery,
     PaddingNode,
     SequenceNode,
     StringNode,
 )
 from snuba.clickhouse.query import Query
+from snuba.query import Query as AbstractQuery
 from snuba.query.composite import CompositeQuery
 from snuba.query.data_source import DataSource
 from snuba.query.data_source.join import IndividualNode, JoinClause, JoinVisitor
 from snuba.query.data_source.simple import Table
+from snuba.query.expressions import Expression
 from snuba.query.parsing import ParsingContext
 from snuba.request.request_settings import RequestSettings
 
+FormattableQuery = Union[Query, CompositeQuery[Table]]
 
-def format_query(query: Query, settings: RequestSettings) -> FormattedQuery:
+
+def format_query(query: FormattableQuery, settings: RequestSettings) -> FormattedQuery:
     """
-    Formats a Clickhouse Query.
+    Formats a Clickhouse Query from the AST representation into an
+    intermediate structure that can either be serialized into a string
+    (for clickhouse) or extracted as a sequence (for logging and tracing).
+
+    This is the entry point for any type of query, whether simple or
+    composite.
 
     TODO: Remove this method entirely and move the sampling logic
     into a query processor.
     """
 
-    if settings.get_turbo() and not query.get_from_clause().sampling_rate:
-        query.set_from_clause(
-            replace(
-                query.get_from_clause(), sampling_rate=snuba_settings.TURBO_SAMPLE_RATE
+    if isinstance(query, Query):
+        if settings.get_turbo() and not query.get_from_clause().sampling_rate:
+            query.set_from_clause(
+                replace(
+                    query.get_from_clause(),
+                    sampling_rate=snuba_settings.TURBO_SAMPLE_RATE,
+                )
             )
-        )
-
-    return format_simple_query(query)
+        return FormattedQuery(_format_single_query_content(query))
+    else:
+        return FormattedQuery(_format_composite_query_content(query))
```

Comment:
I am building the formatted query object here instead of just relying on format_data_source because the type would be wrong (format_data_source must return FormattedNode, while the output must be FormattedQuery).
Also formatting a top level query is slightly different than formatting a subqueries (we add parentheses in the subqueries). 
So, by relegating format_data_source to the task of formatting the FROM clause (not the top level query), we can make such method always format a subquery when it deals with a query and not care about the difference between that and the top level query.